mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassin.java

```diff
@@ -75,34 +75,33 @@ public void learnHam(Flux<InputStream> messages, Username username) {
     @Override
     @PreDestroy
     public void close() {
-        hamHandle.ifPresent(Disposable::dispose);
-        spamHandle.ifPresent(Disposable::dispose);
+        handle.ifPresent(Disposable::dispose);
     }
 
-    private void createListeners(SpamAssassinConfiguration spamAssassinConfiguration) {
-        if (spamAssassinConfiguration.isEnable()) {
-            Host host = spamAssassinConfiguration.getHost().get();
+    private void createListeners(SpamAssassinConfiguration configuration) {
+        if (configuration.isEnable()) {
+            Host host = configuration.getHost().get();
             SpamAssassinInvoker invoker = new SpamAssassinInvoker(metricFactory, host.getHostName(), host.getPort());
 
-            UnicastProcessor<Tuple2<InputStream, Username>> ham = UnicastProcessor.create();
-            hamQueue = Optional.of(ham);
-            hamHandle = Optional.of(ham
-                .doOnNext(Throwing.consumer(messageUsername -> invoker.learnAsHam(messageUsername.getT1(), messageUsername.getT2())))
-                .subscribeOn(Schedulers.elastic())
-                .subscribe());
-
-            UnicastProcessor<Tuple2<InputStream, Username>> spam = UnicastProcessor.create();
-            spamQueue = Optional.of(spam);
-            spamHandle = Optional.of(spam
-                .doOnNext(Throwing.consumer(messageUsername -> invoker.learnAsHam(messageUsername.getT1(), messageUsername.getT2())))
-                .subscribeOn(Schedulers.elastic())
+            UnicastProcessor<Tuple3<InputStream, Username, Kind>> processor = UnicastProcessor.create(Queues.<Tuple3<InputStream, Username, Kind>>one().get());
+            queue = Optional.of(processor);
+            handle = Optional.of(processor
+                .parallel(configuration.getParallelism().get(), 1)
```

Comment:
we don't need parallel IMO, we can use `flatMap` with the right prefetch value (as far as we call `subscribeOn` on the internal `Mono` in the `flatMap`)

Reply:
It's a blocking call right now, I will change it in the @remk PR

Reply:
I don't get it: can't it simply be wrapped into a `Mono.fromCallable`?

Reply:
Got it, thanks.